packages/terminal/src/browser/search/terminal-search-widget.tsx

```diff
@@ -0,0 +1,173 @@
+/********************************************************************************
+ * Copyright (C) 2019 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from 'inversify';
+import { ReactWidget } from '@theia/core/lib/browser/widgets/react-widget';
+import * as React from 'react';
+import '../../../src/browser/search/terminal-search.css';
+import { Terminal } from 'xterm';
+import * as ReactDOM from 'react-dom';
+import { findNext, findPrevious } from 'xterm/lib/addons/search/search';
+import { ISearchOptions } from 'xterm/lib/addons/search/Interfaces';
+import { Key } from '@theia/core/lib/browser';
+
+export const TerminalSearchWidgetFactory = Symbol('TerminalSearchWidgetFactory');
+export type TerminalSearchWidgetFactory = (terminal: Terminal, node: Element, terminalWdgId: string) => TerminalSearchWidget;
```

Comment:
uses options pattern for extensibility, i.e. pass a single argument of an object type